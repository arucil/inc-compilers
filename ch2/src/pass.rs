pub mod assign_home;
pub mod explicate_control;
pub mod instruction_selection;
pub mod partial_evaluation;
pub mod patch_instructions;
pub mod remove_complex_operands;
pub mod uniquify;
pub mod perilogue;