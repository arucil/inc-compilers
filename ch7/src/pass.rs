pub mod parameter;
mod remove_complex_operands;
mod typecheck;
mod uniquify;
