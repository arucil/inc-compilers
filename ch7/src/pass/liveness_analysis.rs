use asm::{Fun, Label, Program};
use ast::IdxVar;
use ch3::location_set::LocationSet;
use ch4::pass::instruction_selection::Info as OldInfo;
use ch4::pass::liveness_analysis::Info as NewInfo;
use ch5::pass::liveness_analysis::analyze_body;
use std::collections::HashMap;

/// `toplevel_label_live` and `fun_label_live` are maps from labels to sets of
/// live locations before the first instruction of the blocks.
/// 
/// `fun_label_live` is the default live-before sets for all functions.
pub fn analyze_liveness(
  prog: Program<OldInfo, IdxVar>,
  toplevel_label_live: HashMap<Label, LocationSet>,
  fun_label_live: HashMap<Label, LocationSet>,
) -> Program<NewInfo, IdxVar> {
  let funs = prog
    .funs
    .into_iter()
    .map(|fun| Fun {
      info: analyze_body(
        fun.info,
        &fun.blocks,
        fun_label_live.clone(),
      ),
      ..fun
    })
    .collect();
  let info = analyze_body(prog.info, &prog.blocks, toplevel_label_live);
  Program { info, funs, ..prog }
}
