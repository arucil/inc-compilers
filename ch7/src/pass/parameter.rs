use ast::{Error, Exp, ExpKind, FunDef, IdxVar, Program, Type};
use std::collections::HashMap;
use support::Range;

pub fn limit_arity(prog: Program<IdxVar, Type>) -> Program<IdxVar, Type> {
  let mut state = State {
    env: HashMap::new(),
  };
  Program {
    fun_defs: prog
      .fun_defs
      .into_iter()
      .map(|(name, fun)| (name, state.fun_limit(fun)))
      .collect(),
    body: prog
      .body
      .into_iter()
      .map(|exp| state.exp_limit(exp))
      .collect(),
    ..prog
  }
}

struct State {
  env: HashMap<IdxVar, Param>,
}

struct Param {
  vec_type: Type,
  index: u32,
}

impl State {
  fn fun_limit(&mut self, fun: FunDef<IdxVar, Type>) -> FunDef<IdxVar, Type> {
    let mut params = fun.params;
    if params.len() > 6 {
      let vec_type =
        Type::Tuple(params.iter().skip(5).map(|p| p.1.clone()).collect());
      for (i, (var, _)) in params.drain(5..).enumerate() {
        self.env.insert(
          var,
          Param {
            vec_type: vec_type.clone(),
            index: i as u32,
          },
        );
      }
      params.push((
        IdxVar {
          name: "(arity-tmp)".to_owned(),
          index: 0,
        },
        vec_type,
      ));
    }
    let body = self.exp_limit(fun.body);
    self.env.clear();
    FunDef {
      params,
      body,
      ..fun
    }
  }

  fn exp_limit(&mut self, exp: Exp<IdxVar, Type>) -> Exp<IdxVar, Type> {
    match exp.kind {
      ExpKind::Int(_)
      | ExpKind::Str(_)
      | ExpKind::Bool(_)
      | ExpKind::Void
      | ExpKind::NewLine => exp,
      ExpKind::Var(var) | ExpKind::Get(var) => {
        if let Some(param) = self.env.get(&var) {
          make_param_ref(exp.range, exp.ty, param, None)
        } else {
          Exp {
            kind: ExpKind::Var(var),
            ..exp
          }
        }
      }
      ExpKind::Prim { op, args } => Exp {
        kind: ExpKind::Prim {
          op,
          args: args.into_iter().map(|exp| self.exp_limit(exp)).collect(),
        },
        ..exp
      },
      ExpKind::Apply {
        fun,
        args,
        r#struct,
      } => {
        if args.len() > 6 {
          let mut args: Vec<_> =
            args.into_iter().map(|exp| self.exp_limit(exp)).collect();
          let vec: Vec<_> = args.drain(5..).collect();
          let vec_range =
            (vec[0].range.start, vec.last().unwrap().range.end).into();
          let last = Exp {
            ty: Type::Tuple(vec.iter().map(|v| v.ty.clone()).collect()),
            kind: ExpKind::Prim {
              op: (vec_range, "vector"),
              args: vec,
            },
            range: vec_range,
          };
          args.push(last);
          Exp {
            kind: ExpKind::Apply {
              fun: box self.exp_limit(*fun),
              r#struct,
              args,
            },
            ..exp
          }
        } else {
          Exp {
            kind: ExpKind::Apply {
              fun: box self.exp_limit(*fun),
              args: args.into_iter().map(|exp| self.exp_limit(exp)).collect(),
              r#struct,
            },
            ..exp
          }
        }
      }
      ExpKind::Let { var, init, body } => Exp {
        kind: ExpKind::Let {
          var,
          init: box self.exp_limit(*init),
          body: box self.exp_limit(*body),
        },
        ..exp
      },
      ExpKind::If { cond, conseq, alt } => Exp {
        kind: ExpKind::If {
          cond: box self.exp_limit(*cond),
          conseq: box self.exp_limit(*conseq),
          alt: box self.exp_limit(*alt),
        },
        ..exp
      },
      ExpKind::Set { var, exp: exp1 } => {
        let val = self.exp_limit(*exp1);
        if let Some(param) = self.env.get(&var.1) {
          make_param_ref(exp.range, exp.ty, param, Some(val))
        } else {
          Exp {
            kind: ExpKind::Set { var, exp: box val },
            ..exp
          }
        }
      }
      ExpKind::Begin { seq, last } => Exp {
        kind: ExpKind::Begin {
          seq: seq.into_iter().map(|exp| self.exp_limit(exp)).collect(),
          last: box self.exp_limit(*last),
        },
        ..exp
      },
      ExpKind::While { cond, body } => Exp {
        kind: ExpKind::While {
          cond: box self.exp_limit(*cond),
          body: box self.exp_limit(*body),
        },
        ..exp
      },
      ExpKind::Print(args) => Exp {
        kind: ExpKind::Print(
          args.into_iter().map(|exp| self.exp_limit(exp)).collect(),
        ),
        ..exp
      },
      ExpKind::Error(Error::Length(arg)) => Exp {
        kind: ExpKind::Error(Error::Length(box self.exp_limit(*arg))),
        ..exp
      },
      ExpKind::Error(Error::OutOfBounds { index, len }) => Exp {
        kind: ExpKind::Error(Error::OutOfBounds {
          index: box self.exp_limit(*index),
          len: box self.exp_limit(*len),
        }),
        ..exp
      },
      ExpKind::Error(Error::DivByZero) => exp,
      ExpKind::FunRef { .. } => exp,
    }
  }
}

fn make_param_ref(
  range: Range,
  ty: Type,
  param: &Param,
  set_value: Option<Exp<IdxVar, Type>>,
) -> Exp<IdxVar, Type> {
  let name = if set_value.is_some() {
    "vector-set!"
  } else {
    "vector-ref"
  };
  let mut args = vec![
    Exp {
      kind: ExpKind::Var(IdxVar {
        name: "(arity-tmp)".to_owned(),
        index: 0,
      }),
      range,
      ty: param.vec_type.clone(),
    },
    Exp {
      kind: ExpKind::Int(param.index as i64),
      range,
      ty: Type::Int,
    },
  ];
  if let Some(v) = set_value {
    args.push(v);
  }
  Exp {
    kind: ExpKind::Prim {
      op: (range, name),
      args,
    },
    range,
    ty,
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn arity() {
    let prog = parse(
      r#"
(define (foo [a Int] [b Bool]) : Void
  (print (bar a))
  (set! b (not b))
  (qux 1 2 3 4 5 6)
  (baz 1 "a" (vector) (void) b (* a 2) (void) (not b)))
(define (bar [x Int]) : Bool
  (> x 0))
(define (qux [a Int] [b Int] [c Int] [d Int] [e Int] [f Int]) : Bool
  (> a (+ b c)))
(define (baz [a Int] [b Str] [c ()] [d Void] [e Bool] [f Int] [g Void] [h Bool]) : Void
  (foo (+ a 1) (not h))
  (set! g (void))
  (print b))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = limit_arity(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
