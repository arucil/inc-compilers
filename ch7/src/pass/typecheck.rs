#[cfg(test)]
mod tests {
  use ast::*;
  use ch5::pass::typecheck::typecheck;
  use support::{CompileError, Range};

  #[test]
  fn fun_def() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  (let ([x (+ x 1)])
    (* x y)))
(define (bar [x Int] [y Int]) : Int
  x)
(let ([x (if #t foo bar)])
  (print (x 1 (read))))
      "#,
    )
    .unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Ok(())
    );
  }

  #[test]
  fn ret_type_error() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  "a")
(define (bar [x Int] [y Int]) : Int
  x)
(let ([x (if #t foo bar)])
  (print (x 1 (read))))
      "#,
    )
    .unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 1, end: 43 },
        message: "expected return type Int, found Str".to_owned(),
      })
    );
  }

  #[test]
  fn arity_mistmach() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  0)
(define (bar [x Int] [y Int]) : Int
  x)
(let ([x (if #t foo bar)])
  (print (x 1)))
      "#,
    )
    .unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 119, end: 124 },
        message: "expected 2 argument(s), found 1".to_owned(),
      })
    );
  }

  #[test]
  fn arg_type_error() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  0)
(define (bar [x Int] [y Int]) : Int
  x)
(let ([x (if #t foo bar)])
  (print (x 1 x)))
      "#,
    )
    .unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 119, end: 126 },
        message: "type mismatch for argument 2".to_owned(),
      })
    );
  }

  #[test]
  fn set_global_fun() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  x)
(define (bar [x Int] [y Int]) : Int
  x)
(let ([x (if #t foo bar)])
  (set! x foo)
  (set! bar foo))
      "#,
    )
    .unwrap();
    let result = typecheck(prog);
    assert_eq!(
      result.map(|_| ()),
      Err(CompileError {
        range: Range { start: 127, end: 141 },
        message: "global function bar is immutable".to_owned(),
      })
    );
  }
}
