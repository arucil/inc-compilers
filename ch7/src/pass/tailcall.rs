use asm::{Block, Instr, Program};
use ch6::pass::perilogue;
use ch6::pass::register_allocation::Info;

pub fn patch_tailjmps(mut prog: Program<Info>) -> Program<Info> {
  prog.blocks = patch_body(&prog.info, prog.blocks);
  for fun in &mut prog.funs {
    fun.blocks = patch_body(&fun.info, std::mem::take(&mut fun.blocks));
  }
  prog
}

fn patch_body(info: &Info, body: Vec<Block>) -> Vec<Block> {
  body
    .into_iter()
    .map(|block| patch_block(info, block))
    .collect()
}

fn patch_block(info: &Info, block: Block) -> Block {
  let mut code = vec![];
  for instr in block.code {
    if let Instr::TailJmp {
      label,
      arity: _,
    } = instr
    {
      code.extend(perilogue::make_epilogue(info).code);
      code.push(Instr::Jmp(label));
    } else {
      code.push(instr);
    }
  }
  Block {
    label: block.label,
    code,
  }
}
