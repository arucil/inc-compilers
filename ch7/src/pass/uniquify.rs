#[cfg(test)]
mod tests {
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn params() {
    let prog = parse(
      r#"
(define (foo [x Int] [y Int]) : Int
  (+ x y))
(define (bar [x Int] [y Int]) : Int
  (let ([x (+ x 1)] [f bar])
    (set! f foo)
    (* x y)))
(let ([x (if #t foo bar)])
  (print (x 1 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let result = uniquify::uniquify(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
