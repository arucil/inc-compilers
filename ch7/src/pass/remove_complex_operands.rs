#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn fun_ref() {
    let prog = ast::parse(
      r#"
(define (foo [f (Int -> Int)]) : Void
  (f 1)
  (void))
(define (bar [x Int]) : Int
  (+ (bar x) 1))
(let ([x bar])
  (foo bar)
  (foo x))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn complex() {
    let prog = ast::parse(
      r#"
(define (foo [x (array-of (array-of Int))]) : Void
  (let ([y (read)] [z y])
    (vector-set! (vector-ref x (remainder (quotient 7 y) 2)) 3 z)
    (and #t (> z 1))
    (set! z (+ z 1))))
(define-type my-str
  (struct
    [s Str]))
(define (bar [s my-str]) : Void
  (set-my-str-s! s (string-append "-" (string-append "," (my-str-s s)))))
(void)
      "#,
    )
    .unwrap();
    let prog = ch5::pass::typecheck::typecheck(prog).unwrap();
    let prog = ch6::pass::r#struct::desugar_struct(prog);
    let prog = ch4::pass::shrink::shrink(prog);
    let prog = ch4::pass::uniquify::uniquify(prog);
    let prog = ch5::pass::uncover_get::uncover_get(prog);
    let prog = ch6::pass::array_bounds::insert_bounds_check(prog);
    let prog = ch6::pass::division::insert_division_check(prog);
    let prog = ch6::pass::string::expose_string_concat(prog);
    let prog =
      ch2::pass::remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(prog.to_string_pretty());
  }
}
