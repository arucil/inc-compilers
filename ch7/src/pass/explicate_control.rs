#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use ch6::pass::array_bounds;
  use insta::assert_snapshot;

  #[test]
  fn call_in_cond() {
    let prog = ast::parse(
      r#"
(define (foo [x (Int Bool Void)]) : Bool
  (if (vector-ref x 1)
    (not (vector-ref x 1))
    (> 3 2)))
(print
  (if (foo (vector (read) #t (void)))
    (begin
      (print 3)
      #t)
    #f))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn call_effect() {
    let prog = ast::parse(
      r#"
(define (foo [x (Int Bool Void)]) : Bool
  (if (vector-ref x 1)
    (not (vector-ref x 1))
    (> 3 2)))
(foo (vector 1 (foo (vector 1 #f (void))) (newline)))
(void)
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = array_bounds::insert_bounds_check(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn void_in_param() {
    let prog = ast::parse(
      r#"
(define (foo [x Int] [y Void] [z Str]) : Bool
  (set! y (void))
  (let ([t (void)])
    (set! t y))
  y
  (print z)
  (not (eq? x 3)))
(void)
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = array_bounds::insert_bounds_check(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn tailcall() {
    let prog = ast::parse(
      r#"
(define (foo [x Int] [y Void] [z Str]) : Bool
  (set! y (void))
  (let ([t (void)])
    (set! t y))
  y
  (print z)
  (foo (+ x 1) y "z"))
(void)
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = array_bounds::insert_bounds_check(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
