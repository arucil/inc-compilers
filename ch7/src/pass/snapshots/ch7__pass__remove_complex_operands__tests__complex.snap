---
source: ch7/src/pass/remove_complex_operands.rs
expression: prog.to_string_pretty()
---
(define
 (foo [x.0 (array-of (array-of Int))]) : Void
 (let
  [y.1 (read)]
  (let
   [z.2 y.1]
   (begin
    (let
     [(bounds-tmp).2
      (let
       [(bounds-tmp).1
        (let
         [(div-tmp).1
          (if (eq? y.1 0)
           (div-by-zero)
           (quotient 7 y.1))]
         (if (eq? 2 0)
          (div-by-zero)
          (remainder (div-tmp).1 2)))]
       (if
        (if (>= (bounds-tmp).1 0)
         (let
          [(tmp).0 (vector-length x.0)]
          (< (bounds-tmp).1 (tmp).0))
         #f)
        (vector-ref x.0 (bounds-tmp).1)
        (let
         [(tmp).1 (vector-length x.0)]
         (out-of-bounds-error
          (bounds-tmp).1
          (tmp).1))))]
     (if
      (if (>= 3 0)
       (let
        [(tmp).2 (vector-length (bounds-tmp).2)]
        (< 3 (tmp).2))
       #f)
      (let
       [(tmp).3 {z.2}]
       (vector-set! (bounds-tmp).2 3 (tmp).3))
      (let
       [(tmp).4 (vector-length (bounds-tmp).2)]
       (out-of-bounds-error 3 (tmp).4))))
    (if #t
     (let
      [(tmp).5 {z.2}]
      (> (tmp).5 1))
     #f)
    (set!
     z.2
     (let
      [(tmp).6 {z.2}]
      (+ (tmp).6 1)))))))
(define (bar [s.3 Id { idx: 0 }]) : Void
 (let
  [(str-tmp).5 "-"]
  (let
   [(str-tmp).6
    (let
     [(str-tmp).1 ","]
     (let
      [(str-tmp).2 (vector-ref s.3 0)]
      (let
       [(str-tmp).3 (string-length (str-tmp).1)]
       (let
        [(str-tmp).4
         (let
          [(tmp).7 (string-length (str-tmp).2)]
          (let
           [(tmp).8 (+ (str-tmp).3 (tmp).7)]
           (alloc-string (tmp).8)))]
        (begin
         (copy-string! (str-tmp).4 0 (str-tmp).1)
         (copy-string!
          (str-tmp).4
          (str-tmp).3
          (str-tmp).2)
         (str-tmp).4)))))]
   (let
    [(str-tmp).7 (string-length (str-tmp).5)]
    (let
     [(str-tmp).8
      (let
       [(tmp).9 (string-length (str-tmp).6)]
       (let
        [(tmp).10 (+ (str-tmp).7 (tmp).9)]
        (alloc-string (tmp).10)))]
     (let
      [(tmp).11
       (begin
        (copy-string! (str-tmp).8 0 (str-tmp).5)
        (copy-string!
         (str-tmp).8
         (str-tmp).7
         (str-tmp).6)
        (str-tmp).8)]
      (vector-set! s.3 0 (tmp).11)))))))
(void)
